File: 2630.py
# project : baekjoon python algorithm practice
# user : marble
# date : 20. 3. 19.
# time : 오후 9:21
# problem : https://www.acmicpc.net/problem/2630
# problem type : Divide and Conquer

# 분할 정복으로 깔끔하게 풀어낸 문제
# 4분할로 나누어서 부분문제를 풀고, 4개 영역으로 종이 수를 센다.

def f(start_x, start_y, length):
    global white, blue
    if length == 1: # 길이 1일때 리턴
        return arr[start_x][start_y]

    areas = []
    areas.append(f(start_x, start_y, length // 2))
    areas.append(f(start_x + length // 2, start_y, length // 2))
    areas.append(f(start_x, start_y + length // 2, length // 2))
    areas.append(f(start_x + length // 2, start_y + length //2, length // 2))

    sum = areas[0] + areas[1] + areas[2] + areas[3] # 모든 파란 색종이의 수
    if sum != length ** 2 and sum != 0: # 파랑과 흰색이 섞여있으면
        for i in range(4):
            if areas[i] == (length // 2) ** 2: # 한 부분이 전부 파랑이면 파란 종이 ++
                blue += 1
            elif areas[i] == 0: # 한 부분이 전부 하양이면 흰 종이 ++
                white += 1
    return sum

import sys
sys.setrecursionlimit(10 ** 6)

n = int(sys.stdin.readline())
arr = []
for i in range(n):
    arr.append(list(map(int, sys.stdin.readline().split())))

white, blue = 0, 0
f(0, 0, n)
print(white)
print(blue)

File: 2630.cpp
//
// Created by 이지호 on 21. 3. 18..
//

#include <iostream>
#include <cmath>
#include <algorithm>
#include <functional>
#include <stack>
#include <vector>
#include <queue>
#include <map>
#include <set>
#include <string>
#include <string.h>

using namespace std;
int n;
int arr[200][200] = {0};
int white = 0, blue = 0;

int divide(int s_x, int s_y, int size) {
    if(size == 1) {
        return arr[s_x][s_y];
    }

    int a1 = divide(s_x, s_y, size/2);
    int a2 = divide(s_x + size / 2, s_y, size/2);
    int a3 = divide(s_x, s_y+size/2, size/2);
    int a4 = divide(s_x+size/2, s_y+size/2, size/2);

    if(a1 == a2 && a2 == a3 && a3 == a4) {
        return a1;
    } else {
        if(a1 == 1) blue++;
        else if(a1 == 0) white++;

        if(a2 == 1) blue++;
        else if(a2 == 0) white++;

        if(a3 == 1) blue++;
        else if(a3 == 0) white++;

        if(a4 == 1) blue++;
        else if(a4 == 0) white++;

        return -1;
    }
}

int main() {
    cin.tie(NULL);
    ios_base::sync_with_stdio(false);

    cin >> n;
    for(int i=0;i<n;i++) {
        for(int j=0;j<n;j++) {
            cin >> arr[i][j];
        }
    }

    int x = divide(0, 0, n);
    if(x == 1) blue++;
    else if(x == 0) white++;
    cout << white << "\n" << blue;


    return 0;
}